chainermn/extensions/multi_node_evaluator.py

```diff
@@ -1,40 +1,36 @@
-def create_multi_node_evaluator(actual_evaluator, communicator):
+import types
+
+
+def create_multi_node_evaluator(evaluator, communicator):
     """Create a multi node evaluator from a normal evaluator.
 
+    Actually patches the evaluator to work in multinode
+    environment. This method adds several hidden attributes starting
+    with _mn_ prefix. After the patch, original evaluator does not
+    work correctly in non-MPI environment.
+
     Args:
-        actual_evaluator: evaluator
+        evaluator: evaluator to be patched
             (e.g., ``chainer.training.extensions.Evaluator``)
         communicator: ChainerMN communicator
 
     Returns:
-        The multi node evaluator based on ``actual_evaluator``.
+        The multi node evaluator based on ``evaluator``.
 
     """
 
-    class MultiNodeEvaluator(type(actual_evaluator)):
-
-        def __init__(self, actual_evaluator_, communicator_):
-            if hasattr(communicator_, 'mpi_comm'):
-                communicator_ = communicator_.mpi_comm
-
-            super(MultiNodeEvaluator, self).__setattr__(
-                'communicator', communicator_)
-            super(MultiNodeEvaluator, self).__setattr__(
-                'actual_evaluator', actual_evaluator_)
-
-        def __getattr__(self, attr_name):
-            return getattr(self.actual_evaluator, attr_name)
-
-        def __setattr__(self, attr_name, value):
-            setattr(self.actual_evaluator, attr_name, value)
-
-        def evaluate(self):
-            local_mean_dict = self.actual_evaluator.evaluate()
-            global_mean_dict = {
-                name:
-                    self.communicator.allreduce(value) / self.communicator.size
-                for name, value in sorted(local_mean_dict.items())
-            }
-            return global_mean_dict
-
-    return MultiNodeEvaluator(actual_evaluator, communicator)
+    evaluator._mn_original_evaluate = evaluator.evaluate
+    evaluator._mn_communicator = communicator
+
+    def new_evaluate(self):
+        local_mean_dict = self._mn_original_evaluate()
+        global_mean_dict = {
+            name:
+            self._mn_communicator.mpi_comm.allreduce(
+                value) / self._mn_communicator.size
+            for name, value in sorted(local_mean_dict.items())
+        }
+        return global_mean_dict
+
+    evaluator.evaluate = types.MethodType(new_evaluate, evaluator)
```

Comment:
Seemingly this code works for both Python 2 and 3 now, but I'm a bit conscious of using `types.MethodType`, which is not documented in the official Python 3 document (https://docs.python.org/3/library/types.html#types.MethodType). How about using `six.create_bound_method` (https://pythonhosted.org/six/#six.create_bound_method)

Reply:
I agree that and will replace with six in the next commit, after　we concluded on compatibility discussion.